src/app/CASESessionManager.cpp

```diff
@@ -0,0 +1,208 @@
+}
+
+Transport::PeerAddress CASESessionManager::ToPeerAddress(const Dnssd::ResolvedNodeData & nodeData) const
+{
+    Inet::InterfaceId interfaceId;
+
+    // Only use the mDNS resolution's InterfaceID for addresses that are IPv6 LLA.
+    // For all other addresses, we should rely on the device's routing table to route messages sent.
+    // Forcing messages down an InterfaceId might fail. For example, in bridged networks like Thread,
+    // mDNS advertisements are not usually received on the same interface the peer is reachable on.
+    if (nodeData.mAddress[0].IsIPv6LinkLocal())
+    {
+        interfaceId = nodeData.mInterfaceId;
+    }
+
+    return Transport::PeerAddress::UDP(nodeData.mAddress[0], nodeData.mPort, interfaceId);
+}
+
+void CASESessionManager::OnNodeIdResolved(const Dnssd::ResolvedNodeData & nodeData)
+{
+    VerifyOrReturn(mInitialized);
+
+    if (mInitParams.dnsCache != nullptr)
+    {
+        LogErrorOnFailure(mInitParams.dnsCache->Insert(nodeData));
+    }
+
+    OperationalDeviceProxy * session = FindSession(nodeData.mPeerId.GetNodeId());
+    VerifyOrReturn(session != nullptr);
+
+    LogErrorOnFailure(session->UpdateDeviceData(
+        ToPeerAddress(nodeData), nodeData.GetMrpRetryIntervalIdle().ValueOr(CHIP_CONFIG_MRP_DEFAULT_IDLE_RETRY_INTERVAL),
+        nodeData.GetMrpRetryIntervalActive().ValueOr(CHIP_CONFIG_MRP_DEFAULT_ACTIVE_RETRY_INTERVAL)));
+}
+
+void CASESessionManager::OnNodeIdResolutionFailed(const PeerId & peer, CHIP_ERROR error)
+{
+    ChipLogError(Controller, "Error resolving node id: %s", ErrorStr(error));
+}
+
+CHIP_ERROR CASESessionManager::GetDeviceAddressAndPort(NodeId deviceId, Inet::IPAddress & addr, uint16_t & port)
+{
+    VerifyOrReturnError(mInitialized, CHIP_ERROR_INCORRECT_STATE);
+
+    OperationalDeviceProxy * session = FindSession(deviceId);
+    VerifyOrReturnError(session != nullptr, CHIP_ERROR_NOT_CONNECTED);
+    VerifyOrReturnError(session->GetAddress(addr, port), CHIP_ERROR_NOT_CONNECTED);
+    return CHIP_NO_ERROR;
+}
+
+void CASESessionManager::OnNewConnection(SessionHandle sessionHandle, Messaging::ExchangeManager * mgr) {}
+
+void CASESessionManager::OnConnectionExpired(SessionHandle sessionHandle, Messaging::ExchangeManager * mgr)
+{
+    VerifyOrReturn(mInitialized);
+
+    OperationalDeviceProxy * session = FindSession(sessionHandle);
+    VerifyOrReturn(session != nullptr,
+                   ChipLogDetail(Controller, "OnConnectionExpired was called for unknown device, ignoring it."));
+
+    session->OnConnectionExpired(sessionHandle);
+}
+
+CHIP_ERROR CASESessionManager::OnMessageReceived(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
+                                                 System::PacketBufferHandle && msgBuf)
+{
+    OperationalDeviceProxy * session = nullptr;
+
+    VerifyOrExit(mInitialized, ChipLogError(Controller, "OnMessageReceived was called in incorrect state"));
```

Comment:
could we not use goto exit patters, especially if the exit side has no cleanup part?